miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultFileAttachmentService.java

```diff
@@ -0,0 +1,197 @@
+    while (targetFile.exists()) {
+      saveFilename = Long.toString(Long.valueOf(saveFilename) + 1);
+      targetFile = new File(saveDir, saveFilename);
+    }
+    if (LimsUtils.checkDirectory(saveDir, true)) {
+      file.transferTo(targetFile);
+    } else {
+      throw new IOException("Cannot upload file - check that the directory specified in miso.properties exists and is writable");
+    }
+
+    try {
+      FileAttachment attachment = new FileAttachment();
+      attachment.setFilename(file.getOriginalFilename());
+      attachment.setPath(relativeDir + File.separator + saveFilename);
+      attachment.setCreator(authorizationManager.getCurrentUser());
+      attachment.setCreationTime(new Date());
+      managed.getAttachments().add(attachment);
+      attachableStore.save(managed);
+    } catch (Exception e) {
+      targetFile.delete();
+      throw e;
+    }
+  }
+
+  @Override
+  public void delete(Attachable object, FileAttachment attachment) throws IOException {
+    if (!object.getAttachments().contains(attachment)) {
+      throw new IllegalArgumentException("Attachment does not belong to this object");
+    }
+    Attachable managed = attachableStore.getManaged(object);
+    FileAttachment managedAttachment = managed.getAttachments().stream()
+        .filter(a -> a.getId() == attachment.getId())
+        .findFirst().orElseThrow(() -> new IllegalArgumentException("Attachment not found in persisted entity"));
+
+    authorizationManager.throwIfNonAdminOrMatchingOwner(managedAttachment.getCreator());
+
+    File file = new File(makeFullPath(managedAttachment.getPath()));
+    if (file.exists()) {
+      if (!file.isFile()) {
+        throw new IOException("Cannot delete. Not a file");
+      }
+      if (!file.canWrite()) {
+        throw new IOException("Cannot delete file. Permission denied");
+      }
+    }
+
+    managed.getAttachments().remove(attachment);
+    attachableStore.save(managed);
+
+    if (file.exists()) {
+      deleteFileOrLog(file, managed, managedAttachment);
+    }
+  }
+
+  @Override
+  public void afterDelete(Attachable object) {
+    for (FileAttachment a : object.getAttachments()) {
+      File file = new File(makeFullPath(a.getPath()));
+      deleteFileOrLog(file, object, a);
+    }
+    File dir = new File(makeFullPath(makeRelativeDir(object.getAttachmentsTarget(), object.getId())));
+    if (dir.exists() && dir.isDirectory()) {
+      if (dir.listFiles() != null) {
+        for (File file : dir.listFiles()) {
+          deleteFileOrLog(file, object, null);
+        }
+      }
+      if (!dir.delete()) {
+        log.error("Failed to delete directory for deleted object: " + dir.getAbsolutePath());
+      }
+    }
+  }
+
+  private void deleteFileOrLog(File file, Attachable object, FileAttachment attachment) {
+    if (!file.delete()) {
+      log.error(String.format("File no longer associated with an object, but failed to delete: %s%s from %s %d", file.getName(),
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') An 'int' is expected rather than a String. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2275)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Invoke method(s) only conditionally. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2629)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') 5th argument is not used. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)